Remove shared IL stubs for P/Invokes.

AaronRobinsonMSFT opened the pull request:
This removes usage of the secret argument in P/Invokes. It removes sharing for all P/Invoke IL stubs except CALLI.

Depends on https://github.com/dotnet/runtime/pull/118462

AaronRobinsonMSFT committed:
Remove shared IL stubs for P/Invokes.

This removes usage of the secret argument in P/Invokes.
It also removes all usage of shared IL stubs.

AaronRobinsonMSFT commented:
> There is more cleanup that should be done as part of this:
> 
> * MethodDesc::RequiresMethodDescCallingConvention should drop `IsNDirect` condition
> * `CORJIT_FLAG_PUBLISH_SECRET_PARAM` should not be set for PInvoke stubs
> * `HasMDContextArg()` should drop `IsNDirect` condition
> * There are probably some changes needed around `GetActualInteropMethodDesc`

Yep. This was a quick and dirty solution I could pass off to the interpreter team.

AaronRobinsonMSFT commented on src/coreclr/vm/dllimport.h:
```diff
@@ -251,28 +251,6 @@ inline bool SF_IsCOMEventCallStub      (DWORD dwStubFlags) { LIMITED_METHOD_CONT
 inline bool SF_IsFieldGetterStub       (DWORD dwStubFlags) { LIMITED_METHOD_CONTRACT; return COM_ONLY(dwStubFlags < NDIRECTSTUB_FL_INVALID && 0 != (dwStubFlags & NDIRECTSTUB_FL_FIELDGETTER)); }
 inline bool SF_IsFieldSetterStub       (DWORD dwStubFlags) { LIMITED_METHOD_CONTRACT; return COM_ONLY(dwStubFlags < NDIRECTSTUB_FL_INVALID && 0 != (dwStubFlags & NDIRECTSTUB_FL_FIELDSETTER)); }
 
-inline bool SF_IsSharedStub(DWORD dwStubFlags)
```
I see this is done for getters and setters, but I'm not sure how valuable those really are. Since we are pushing source generated COM, I think the same argument applied with P/Invoke and stub sharing is the same as COM.

jkotas replied:
> I see this is done for getters and setters

I am not following. COM interop stubs should be shared for all ordinary methods (including property getters and property setters) today. 

I would avoid touching how built-in COM interop works. It is very likely to have fallout. For example, do we know how much of built-in COM interop is used by VS side-car processes? If we stop sharing COM interop stubs, are VS perf gates that count number of methods JITed going to fail?

AaronRobinsonMSFT commented:
> * MethodDesc::RequiresMethodDescCallingConvention should drop `IsNDirect` condition

Done. 

> * `CORJIT_FLAG_PUBLISH_SECRET_PARAM` should not be set for PInvoke stubs

Inserted a `SF_IsForwardStub(m_dwStubFlags)` check before the flag is set in stubs.

> * `HasMDContextArg()` should drop `IsNDirect` condition

Removed the PInvoke check.

> * There are probably some changes needed around `GetActualInteropMethodDesc`

Usage is blocked by `HasMDContextArg()` so updates flow to the single usage. Suggestions on what I could assert?

AaronRobinsonMSFT committed:
Feedback
Clean-up dead code in RequiresMethodDescCallingConvention and RequiresStableEntryPoint.

jkotas commented:
> Usage is blocked by HasMDContextArg() so updates flow to the single usage. 

If I am reading the code correctly, https://github.com/dotnet/runtime/blob/2db9fdf5b4083c7459fc2150e8589a81ac7ebb39/src/coreclr/vm/stackwalk.cpp#L3030 will cause stackwalker to report the IL stub method desc instead of the actual method desc. It may have undesirable fallout - I am thinking about problems like:
- Are IL stubs going to show up in stacktraces, e.g. when P/Invoke throws an exception that gets unhandled? 
- Can it confuse debuggers?

AaronRobinsonMSFT committed:
Bring back sharing for COM stubs.

jkotas commented:
I think it would best to minimize stackwalker dependencies if possible.

Can we get `InlinedCallFrame::m_Datum` initialized to point to the actual target method desc? It is done here: https://github.com/dotnet/runtime/blob/97d94229eb62859038cd05fb864fd39e9be0a1a0/src/coreclr/jit/lower.cpp#L6595-L6612 .

It is also going to fix a type mismatch: `InlinedCallFrame::m_Datum is expected to be `NDirectMethodDesc` (modulo some bit twiddling and special cases), but it is DynamicMethod desc with this change.

AaronRobinsonMSFT commented:
> It is also going to fix a type mismatch: InlinedCallFrame::m_Datum is expected to be NDirectMethodDesc (modulo some bit twiddling and special cases), but it is DynamicMethodDesc with this change.

Sure, I can give that a shot. This doesn't ever appear to be set for coreclr JIT, the `embedMethodHandle` is a nop.

https://github.com/dotnet/runtime/blob/97d94229eb62859038cd05fb864fd39e9be0a1a0/src/coreclr/vm/jitinterface.cpp#L10455-L10473

Side note, I've really got no idea what is expected here. All of this is a huge confusion about various states of fields and very loosely typed in practice. I'm really just randomly guessing what things are expected at this point, sorry.

AaronRobinsonMSFT committed:
Set the CodeHeader to contain the NDirectMethodDesc
instead of the IL Stub's DynamicMethodDesc.

AaronRobinsonMSFT committed:
Clean up the CodeHeapRequestInfo type.

jkotas commented on src/coreclr/vm/method.hpp:
```diff
@@ -2859,6 +2858,14 @@ class DynamicMethodDesc : public StoredSigMethodDesc
+            && !HasFlags(FlagIsCALLI | FlagIsDelegate)
+            && GetILStubType() == StubCLRToNativeInterop;
+    }
+    bool IsFPtrToDelegateStub() const
```
Is this name reversed? This stub goes from delegate to fptr. 

(It is not great that we use different names for the same thing. I believe that this is called ForwardDelegateStub in other files.)

jkotas replied:
This method may be unnecessary depending on resolution of the other feedback.

AaronRobinsonMSFT commented on src/coreclr/vm/prestub.cpp:
```diff
@@ -383,7 +383,9 @@ PCODE MethodDesc::PrepareILBasedCode(PrepareCodeConfig* pConfig)
-                if (stubMethodDesc->IsILStub() && stubMethodDesc->IsPInvokeStub())
+                if (stubMethodDesc->IsILStub()
+                    && (stubMethodDesc->IsPInvokeStub()
+                        || stubMethodDesc->IsFPtrToDelegateStub()))
```
> so it is not clear to me why we need this change.

It was changed because I changed the implementation of `IsPInvokeStub()`.

> I think if (GetModule()->IsReadyToRun() && !GetModule()->GetReadyToRunInfo()->HasNonShareablePInvokeStubs()) is always going to be false.

Okay, if that is the case I agree. I'll try to confirm where we are at first.

AaronRobinsonMSFT replied:
I agree about CrossGen. It looks like the `ReadyToRunFlags.READYTORUN_FLAG_NonSharedPInvokeStubs` flag is set unconditionally there. I'll remove this.

AaronRobinsonMSFT committed:
Feedback.
Fix "Forward PInvoke" and CALLI confusion.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into remove_shared_ilstubs

jkotas commented on src/coreclr/vm/dllimport.h:
```diff
@@ -270,20 +281,14 @@ inline bool SF_IsSharedStub(DWORD dwStubFlags)
         return false;
     }
 
+    if (SF_IsForwardPInvokeStub(dwStubFlags))
```
Should this exclude Delegate and Calli like in the other place?

AaronRobinsonMSFT replied:
Yep. I just got back from a run and am in the middle of cleaning that up. Occurred to me while running :)

AaronRobinsonMSFT committed:
Make order checks consistent for stub types.

AaronRobinsonMSFT committed:
Copy/paste error.

AaronRobinsonMSFT commented on src/coreclr/vm/dllimport.cpp:
```diff
@@ -2235,21 +2233,23 @@ DWORD NDirectStubLinker::EmitProfilerBeginTransitionCallback(ILCodeStream* pcsEm
+        // Some other stub without the MD as the secret parameter, so pass null.
         pcsEmit->EmitLoadNullPtr();
     }
 
```
Blech... I convinced myself this was okay - "we don't have it", but of course we do.

AaronRobinsonMSFT committed:
Emit MD in profiler scenarios

AaronRobinsonMSFT committed:
Attempt to avoid GCC assumption on null.

AaronRobinsonMSFT committed:
Try to fix up GCC.

AaronRobinsonMSFT committed:
Initialize locals.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into remove_shared_ilstubs

AaronRobinsonMSFT committed:
Missed edits

AaronRobinsonMSFT committed:
Missed in merge.

AaronRobinsonMSFT committed:
Fix vararg interop

AaronRobinsonMSFT committed:
Update JIT interface

AaronRobinsonMSFT committed:
Add MethodDesc to VASigCookie

Stop passing the MethodDesc to all forward
P/Invoke scenarios.
Address stepping on vararg P/Invokes (mixed mode).

AaronRobinsonMSFT commented on src/coreclr/vm/method.cpp:
```diff
@@ -2476,20 +2476,23 @@ MethodImpl *MethodDesc::GetMethodImpl()
+    if (IsCLRToCOMCall())
         return TRUE;
 
+    // Interop marshalling of vararg needs MethodDesc calling convention
```
Done.

jkotas commented on src/coreclr/vm/stubmgr.cpp:
```diff
@@ -1783,41 +1783,43 @@ BOOL ILStubManager::TraceManager(Thread *thread,
 
         trace->InitForManaged(target);
     }
+    else if (pStubMD->IsPInvokeStub())
```
Can we ever end up with pStubMD->IsPInvokeStub() == true here now that we are patching the MethodDesc in the code header?

AaronRobinsonMSFT committed:
Fix build breaks.

AaronRobinsonMSFT committed:
Remove unused path

AaronRobinsonMSFT commented on src/coreclr/vm/method.hpp:
```diff
@@ -2900,7 +2899,11 @@ class DynamicMethodDesc : public StoredSigMethodDesc
-        return IsCLRToCOMStub() || (IsPInvokeStub() && !HasFlags(FlagIsDelegate));
+        bool needsArg = IsCLRToCOMStub();
+#ifndef DACCESS_COMPILE
+        _ASSERTE(needsArg == !!RequiresMDContextArg());
```
Boo. I also confused the `IsPInvoke()` and `IsPInvokeStub()` previously.

AaronRobinsonMSFT committed:
Remove invalid assert.

AaronRobinsonMSFT committed:
Add cast.

AaronRobinsonMSFT commented on src/coreclr/vm/frames.h:
```diff
@@ -2203,10 +2203,10 @@ class InlinedCallFrame : public Frame
     void UpdateRegDisplay_Impl(const PREGDISPLAY, bool updateFloats = false);
 
-    // m_Datum contains MethodDesc ptr or
+    // m_Datum contains NDirectMethodDesc ptr or
```
```suggestion
    // m_Datum contains PInvokeMethodDesc ptr or
```

AaronRobinsonMSFT commented on src/coreclr/vm/frames.h:
```diff
@@ -2203,10 +2203,10 @@ class InlinedCallFrame : public Frame
     // - on 64 bit host: CALLI target address (if lowest bit is set)
     // - on windows x86 host: argument stack size (if value is <64k)
-    // When m_Datum contains MethodDesc ptr, then on other than windows x86 host
+    // When m_Datum contains NDirectMethodDesc ptr, then on other than windows x86 host
```
```suggestion
    // When m_Datum contains PInvokeMethodDesc ptr, then on other than windows x86 host
```

AaronRobinsonMSFT committed:
Apply suggestions from code review

jkotas commented on src/coreclr/vm/ceeload.cpp:
```diff
@@ -4323,41 +4318,44 @@ VASigCookie *Module::GetVASigCookieWorker(Module* pDefiningModule, Module* pLoad
         {
             CrstHolder ch(&pLoaderModule->m_Crst);
 
+            AllocMemTracker amt;
+
```
```suggestion
 ```

jkotas commented on src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp:
```diff
@@ -5510,7 +5510,7 @@ void MethodContext::repFindCallSiteSig(CORINFO_MODULE_HANDLE  module,
 }
 
-void MethodContext::recGetVarArgsHandle(CORINFO_SIG_INFO* pSig, void** ppIndirection, CORINFO_VARARGS_HANDLE result)
+void MethodContext::recGetVarArgsHandle(CORINFO_SIG_INFO* pSig, CORINFO_METHOD_HANDLE methHnd, void** ppIndirection, CORINFO_VARARGS_HANDLE result)
```
CORINFO_METHOD_HANDLE needs to be added to `key` here and other places under superpmi.

jkotas commented on src/coreclr/tools/Common/JitInterface/ThunkGenerator/ThunkInput.txt:
```diff
@@ -315,7 +315,7 @@ FUNCTIONS
     CORINFO_CLASS_HANDLE getStaticFieldCurrentClass(CORINFO_FIELD_HANDLE    field,  BoolStar pIsSpeculative);
-    CORINFO_VARARGS_HANDLE getVarArgsHandle(CORINFO_SIG_INFO       *pSig, void                  **ppIndirection);
+    CORINFO_VARARGS_HANDLE getVarArgsHandle(CORINFO_SIG_INFO       *pSig, CORINFO_METHOD_HANDLE   methHnd, void                  **ppIndirection);
     bool canGetVarArgsHandle(CORINFO_SIG_INFO       *pSig);
```
`canGetVarArgsHandle` method can be deleted. It is useless - left-over from fragile NGen.

jkotas replied:
Same for `canGetCookieForPInvokeCalliSig`. It can be deleted as well.

(You may want to submit the JIT/EE interface edits as separate prep-work PR.)

AaronRobinsonMSFT committed:
Feedback.

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -6159,7 +6160,8 @@ CORINFO_VARARGS_HANDLE CEEInfo::getVarArgsHandle(CORINFO_SIG_INFO *sig,
 
-    result = CORINFO_VARARGS_HANDLE(module->GetVASigCookie(Signature(sig->pSig, sig->cbSig), &typeContext));
+    MethodDesc* pMD = GetMethod(methHnd);
+    result = CORINFO_VARARGS_HANDLE(module->GetVASigCookie(Signature(sig->pSig, sig->cbSig), pMD, &typeContext));
```
```suggestion
    result = CORINFO_VARARGS_HANDLE(module->GetVASigCookie(Signature(sig->pSig, sig->cbSig), pMD->IsPInvoke() ? pMD : NULL, &typeContext));
```
I think we need to include the methoddesc in the VASigCookie key only if it is a PInvoke. Ordinary varargs should not need it.

AaronRobinsonMSFT committed:
Only embed PInvokeMethodDescs for varargs.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into remove_shared_ilstubs

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into remove_shared_ilstubs

AaronRobinsonMSFT committed:
Merge branch 'main' into remove_shared_ilstubs
